rfc/private-key-handling.md

```diff
@@ -0,0 +1,173 @@
+# RFC: Standardizing Private Key Material Handling
+
+## Summary
+
+This RFC proposes standardizing how we handle private key material across our services, adopting PEM format as the single standard for private key storage and configuration.
+
+## Problem Statement
+
+Our services currently accept private keys in multiple formats:
+
+- PEM files (with inconsistent encryption practices).
+- JSON files (non-standard, service-specific schemas).
+- Base64 encoded strings (raw key material, no metadata).
+- Multibase base64 padded strings (uncommon in standard tooling).
+
+This creates:
+
+- **Security inconsistency**: Different formats have varying security properties.
+- **Implementation complexity**: Each service implements multiple parsers.
+- **Cognitive overhead**: Developers must remember which format works where.
+- **Error-prone deployments**: Format mismatches cause configuration failures.
+
+## Proposed Solution: PEM Standard
+
+**Format**: PEM-encoded private keys using PKCS#8 format
+
+```
+-----BEGIN PRIVATE KEY-----
+MC4CAQA...
+-----END PRIVATE KEY-----
+```
+
+### Why PEM?
+
+1. **Universal tooling support**: OpenSSL, ssh-keygen, and all major crypto libraries support PEM natively.
+2. **Self-documenting format**: Clear headers (`-----BEGIN PRIVATE KEY-----`) identify content type.
+3. **Algorithm agnostic**: Supports our current Ed25519 keys and future algorithms (RSA, ECDSA, etc.).
+4. **Built-in security**: Native support for passphrase encryption.
+5. **Human readable**: Base64 encoded with clear delimiters for debugging.
+6. **Industry standard**: Developers already understand `.pem` files.
+
+### Usage Examples
+
+**Generating Ed25519 keys**:
+
+```bash
+# Generate Ed25519 private key
+openssl genpkey -algorithm ed25519 -out service-key.pem
+
+# View key details
+openssl pkey -in service-key.pem -text -noout
+
+# Extract public key
+openssl pkey -in service-key.pem -pubout -out service-pubkey.pem
+```
+
+**Securing keys**:
+
+```bash
+# Set proper permissions
+chmod 600 service-key.pem
+
+# Encrypt for storage
+openssl pkey -in service-key.pem -out service-key-encrypted.pem -aes256
+
+# Decrypt when needed
+openssl pkey -in service-key-encrypted.pem -out service-key.pem
+```
+
+**Future algorithm support** (no code changes needed):
+
+```bash
+# RSA
+openssl genpkey -algorithm rsa -pkeyopt rsa_keygen_bits:4096 -out rsa-key.pem
+
+# ECDSA
+openssl genpkey -algorithm ec -pkeyopt ec_paramgen_curve:P-256 -out ecdsa-key.pem
+```
+
+## Security Requirements
+
+1. **File permissions**: Private key files MUST have 600 permissions.
+2. **No environment variables**: Private keys MUST NOT be passed via environment variables.
+3. **Encryption at rest**: Keys SHOULD be encrypted when stored.
+4. **Memory handling**: Keys SHOULD be loaded once at startup, not repeatedly read.
+5. **Secure deletion**: Key material SHOULD be securely overwritten when removed from memory.
```

Comment:
We have historically used AWS SSM for deployed services and would continue to do so regardless of encoding (I understand this is not the same situation for storage nodes).